torch/distributed/pipelining/schedules.py

```diff
@@ -2070,6 +2141,294 @@ def backward_stage_index(step):
+                    + chunk * pipeline_parallel_size * num_microbatches
+                    + rank * num_microbatches
+                    + micro
+                )
+
+            count = []
+            for i in range(pipeline_parallel_size):
+                count.append([0] * 6)
+            fbw_mem = [39, -7, -32]
+            max_mem = 39 * (pipeline_parallel_size * 2)
+            end_time = [-1] * n_node
+            cur_time = [0] * pipeline_parallel_size
+            mem = [0] * pipeline_parallel_size
+            rank_bubble = [0] * pipeline_parallel_size
+            pending_w: List[Deque[Tuple[int, int, int]]] = [
+                deque() for _ in range(pipeline_parallel_size)
+            ]
+            schedule: Dict[int, List[Tuple[int, _ComputationType, int]]] = {}
+            for i in range(pipeline_parallel_size):
+                schedule[i] = []
+            stage_str = ["    " * i for i in range(pipeline_parallel_size)]
+            approved_bubble = [-1] * pipeline_parallel_size
+            max_approved_bubble = max(approved_bubble)
+
+            def get_max_rank_bubble(rank=-1):
+                max_rank_bubble = 0
+                for bb in rank_bubble:
+                    max_rank_bubble = max(max_rank_bubble, bb)
+                if rank >= 0:
+                    max_rank_bubble = max(
+                        max_rank_bubble, max_approved_bubble - approved_bubble[rank]
+                    )
+                return max_rank_bubble
+
+            def put_w(rank):
+                assert len(pending_w[rank]) > 0
+                _, chunk_, _ = pending_w[rank].popleft()
+                put(2, chunk_, rank)
+
+            def put(cat, chunk, rank, assert_cnt=True):
+                _tmp = _no_bubble = cur_time[rank] + 1
+                _cnt = count[rank][cat * 2 + chunk]
+                stage_str[rank] += (
+                    "FfBbWw"[cat * 2 + chunk]
+                    + str(_cnt + 1)
+                    + " " * (3 - len(str(_cnt + 1)))
+                )
+                if chunk == 1 and cat < 2:
+                    if rank < pipeline_parallel_size - 1:
+                        _fa_id = get_id(cat, chunk, rank + 1, _cnt)
+                        assert end_time[_fa_id] >= 0
+                        _tmp = max(_tmp, end_time[_fa_id] + 1)
+                if chunk == 0 and cat < 2:
+                    if rank > 0:
+                        _fa_id = get_id(cat, chunk, rank - 1, _cnt)
+                        assert end_time[_fa_id] >= 0, f"{cat}, {chunk}, {rank}, {_cnt}"
+                        _tmp = max(_tmp, end_time[_fa_id] + 1)
+                _id = get_id(cat, chunk, rank, _cnt)
+                if count[rank][0] > 0:
+                    rank_bubble[rank] += _tmp - _no_bubble
+                end_time[_id] = _tmp
+                cur_time[rank] = _tmp
+                mem[rank] += fbw_mem[cat]
+                # noinspection PyTypeChecker
```

Comment:
we actually should not disable type checker, we should add the mypy hints everywhere and make sure its correct. helps catch some bugs.